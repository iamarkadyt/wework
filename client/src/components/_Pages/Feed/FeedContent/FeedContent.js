import React, { Fragment } from 'react'
import Post from '../../../Post/Post'
import CommentsView from '../../../CommentsView/CommentsView'
import { Route } from 'react-router-dom'
import { arrayOf, string } from 'prop-types'
import { postType } from '../../../../types/index'
import FBSpinner from '../../../FBSpinner/FBSpinner'
import NoContent from '../NoContent/NoContent'
import EndOfFeedMessage from '../EndOfFeedMessage/EndOfFeedMessage'

const isEmptyFn = ({ endOfFeed, posts }) => {
  return posts.length === 0 && !!endOfFeed
}

const isLoadingFn = ({ endOfFeed, posts, isLoading }) => {
  return isLoading || (!endOfFeed && posts.length === 0)
}

const isEndOfFeedFn = ({ endOfFeed, isLoading }) => {
  return !!endOfFeed && !isLoading
}

const feedContent = props => {
  const { posts, baseUrl, endOfFeed } = props
  let bottomWidget = null

  if (isEmptyFn(props)) {
    bottomWidget = <NoContent />
  } else if (isLoadingFn(props)) {
    bottomWidget = <FBSpinner />
  } else if (isEndOfFeedFn(props)) {
    bottomWidget = <EndOfFeedMessage endOfFeed={endOfFeed}/>
  }

  return (
    <Fragment>
      <Route path={`${baseUrl}/view-comments/:postId`} render={props => (
          <CommentsView {...props} post={posts.find(
              item => item._id === props.match.params.postId
            )} />
        )} />
      <div className="Feed-body">
        {posts.map(item => <Post key={item._id} {...item} />)}
      </div>
      {bottomWidget}
    </Fragment>
  )
}

feedContent.propTypes = {
  posts: arrayOf(postType).isRequired,
  baseUrl: string.isRequired
}

export { isEmptyFn, isLoadingFn, isEndOfFeedFn }
export default feedContent
