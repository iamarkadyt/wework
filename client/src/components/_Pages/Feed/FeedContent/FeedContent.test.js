import React from 'react'
import FeedContent, { isEmptyFn, isLoadingFn, isEndOfFeedFn } from './FeedContent'
import { BrowserRouter } from 'react-router-dom'
import { shallow } from 'enzyme'
import cloneDeep from 'lodash.clonedeep'

/**
 *  Use following for complete mounts:
 *
 *    import { Provider } from 'react-redux'
 *    import configureMockStore from 'redux-mock-store'
 *
 *    const mockStore = configureMockStore()
 *    const store =  mockStore({
 *      user: { id: '123' },
 *      err: { formErrors: {} }
 *    })
 *
 *    mountedComponent = mount(
 *      <Provider store={store}>
 *        <Router location={props.location} context={{}}>
 *          <CommentsRoute {...props} />
 *        </Router>
 *      </Provider>
 *    )
 */

let mountedComponent, props

const getMockProps = () => {
  return cloneDeep({
    posts: [require('../../../../mocks/posts').mockPost],
    baseUrl: ''
  })
}

const createRouterContext = pathname => {
  return cloneDeep({
    router: {
      history: new BrowserRouter().history,
      route: {
        location: {
          pathname
        },
        match: {},
      }
    }
  })
}

const comp = () => {
  if (!mountedComponent) {
    mountedComponent = shallow(
      <FeedContent {...props} />
    )
  } 
  return mountedComponent
}

describe('FeedContent', () => {
  beforeEach(() => {
    mountedComponent = undefined
    props = getMockProps()
  })

  describe('functions', () => {
    describe('isEndOfFeedFn', () => {
      it('returns true if endOfFeed is defined and isLoading is false', () => {
        const customProps = {
          endOfFeed: 'some error',
          isLoading: false
        }
        expect(isEndOfFeedFn(customProps)).toBe(true)
      })

      it('returns false if endOfFeed is not defined but isLoading is false', () => {
        const customProps = {
          isLoading: false
        }
        expect(isEndOfFeedFn(customProps)).toBe(false)
      })

      it('returns false if endOfFeed is defined but isLoading is true', () => {
        const customProps = {
          endOfFeed: 'some error',
          isLoading: true
        }
        expect(isEndOfFeedFn(customProps)).toBe(false)
      })

      it('returns false if endOfFeed is defined and isLoading is true', () => {
        const customProps = {
          endOfFeed: 'some error',
          isLoading: true
        }
        expect(isEndOfFeedFn(customProps)).toBe(false)
      })
    })

    describe('isLoadingFn', () => {
      it('returns true if isLoading is true', () => {
        const customProps = {
          isLoading: true
        }
        expect(isLoadingFn(customProps)).toBe(true)
      })

      it('returns true if isLoading is false but endOfFeed is not defined and no posts were loaded', () => {
        const customProps = {
          isLoading: false,
          posts: []
        }
        expect(isLoadingFn(customProps)).toBe(true)
      })

      it('returns true if isLoading is true and endOfFeed is not defined and no posts were loaded', () => {
        const customProps = {
          isLoading: true,
          posts: []
        }
        expect(isLoadingFn(customProps)).toBe(true)
      })

      it('returns false if isLoading is false and endOfFeed is defined but no posts were loaded', () => {
        const customProps = {
          isLoading: false,
          posts: [],
          endOfFeed: 'error!'
        }
        expect(isLoadingFn(customProps)).toBe(false)
      })

      it('returns false if isLoading is false and endOfFeed is not defined but some posts were loaded', () => {
        const customProps = {
          isLoading: false,
          posts: [ {}, {}, {} ],
        }
        expect(isLoadingFn(customProps)).toBe(false)
      })

      it('returns false if isLoading is false, endOfFeed is defined and some posts were loaded', () => {
        const customProps = {
          isLoading: false,
          endOfFeed: 'error!',
          posts: [ {}, {}, {} ],
        }
        expect(isLoadingFn(customProps)).toBe(false)
      })
    })

    describe('isEmptyFn', () => {
      it('returns true if no posts were loaded and endOfFeed is defined', () => {
        const customProps = {
          endOfFeed: 'defined!',
          posts: []
        }
        expect(isEmptyFn(customProps)).toBe(true)
      })

      it('returns false if some posts were loaded and endOfFeed is defined', () => {
        const customProps = {
          endOfFeed: 'defined!',
          posts: [ {} ]
        }
        expect(isEmptyFn(customProps)).toBe(false)
      })

      it('returns false if no posts were loaded but endOfFeed is undefined', () => {
        const customProps = {
          posts: []
        }
        expect(isEmptyFn(customProps)).toBe(false)
      })

      it('returns false if some posts were loaded and endOfFeed is undefined', () => {
        const customProps = {
          posts: [ {} ],
          endOfFeed: 'some err'
        }
        expect(isEmptyFn(customProps)).toBe(false)
      })
    })
  })

  it('must generate expected amount of Posts', () => {
    expect(comp().find('withRouter(Connect(Post))')).toHaveLength(props.posts.length)
  })

  describe('snapshots', () => {
    it('Route render-out matches snapshot', () => {
      const context = createRouterContext(`${props.baseUrl}/view-comments/${props.posts[0]._id}`)
      expect(comp().find('Route').dive({ context })).toMatchSnapshot()
    })

    it('matches snapshot', () => {
      expect(comp()).toMatchSnapshot()
    })
  })
})
