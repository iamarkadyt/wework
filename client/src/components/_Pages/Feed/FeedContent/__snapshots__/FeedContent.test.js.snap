// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedContent snapshots Route render-out matches snapshot 1`] = `
<Connect(CommentsView)
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "pathname": "/view-comments/5b9f4728bc40273febf52706",
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {
        "postId": "5b9f4728bc40273febf52706",
      },
      "path": "/view-comments/:postId",
      "url": "/view-comments/5b9f4728bc40273febf52706",
    }
  }
  post={
    Object {
      "__v": 0,
      "_id": "5b9f4728bc40273febf52706",
      "comments": Array [
        Object {
          "_id": "5ba1dd77809c892f580240ef",
          "date": "2018-09-19T05:24:07.399Z",
          "text": "Hahahahahahahaaa!",
          "user": Object {
            "_id": "5b75f023f1957052e52106f5",
            "avatar": "//www.gravatar.com/avatar/254dbff8aa12a023b9f1052ad80b1831?s=200&r=pg&d=mm",
            "name": "John Grapard",
          },
        },
        Object {
          "_id": "5ba1de21f0b9f73054396abf",
          "date": "2018-09-19T05:26:57.632Z",
          "text": "Nuka proverim esli oshibka ostalas...",
          "user": Object {
            "_id": "5b75f023f1957052e52106f5",
            "avatar": "//www.gravatar.com/avatar/254dbff8aa12a023b9f1052ad80b1831?s=200&r=pg&d=mm",
            "name": "John Grapard",
          },
        },
      ],
      "date": "2018-09-17T06:18:16.992Z",
      "likes": Array [
        Object {
          "_id": "5ba1dd61809c892f580240ed",
          "user": "5b75f023f1957052e52106f5",
        },
        Object {
          "_id": "5ba80533058247301e362bf9",
          "user": "5ba1acfc13b542138b29af68",
        },
        Object {
          "_id": "5ba847153c823e00161763b5",
          "user": "5ba83562ed730446b54def34",
        },
      ],
      "text": "Tas nisl commlor sit amet, consec!",
      "user": Object {
        "_id": "5b9eedd621194c102c5280e8",
        "avatar": "//www.gravatar.com/avatar/e13743a7f1db7f4246badd6fd6ff54ff?s=200&r=pg&d=mm",
        "name": "John Doe",
      },
    }
  }
/>
`;

exports[`FeedContent snapshots matches snapshot 1`] = `
<Fragment>
  <Route
    path="/view-comments/:postId"
    render={[Function]}
  />
  <div
    className="Feed-body"
  >
    <withRouter(Connect(Post))
      __v={0}
      _id="5b9f4728bc40273febf52706"
      comments={
        Array [
          Object {
            "_id": "5ba1dd77809c892f580240ef",
            "date": "2018-09-19T05:24:07.399Z",
            "text": "Hahahahahahahaaa!",
            "user": Object {
              "_id": "5b75f023f1957052e52106f5",
              "avatar": "//www.gravatar.com/avatar/254dbff8aa12a023b9f1052ad80b1831?s=200&r=pg&d=mm",
              "name": "John Grapard",
            },
          },
          Object {
            "_id": "5ba1de21f0b9f73054396abf",
            "date": "2018-09-19T05:26:57.632Z",
            "text": "Nuka proverim esli oshibka ostalas...",
            "user": Object {
              "_id": "5b75f023f1957052e52106f5",
              "avatar": "//www.gravatar.com/avatar/254dbff8aa12a023b9f1052ad80b1831?s=200&r=pg&d=mm",
              "name": "John Grapard",
            },
          },
        ]
      }
      date="2018-09-17T06:18:16.992Z"
      key="5b9f4728bc40273febf52706"
      likes={
        Array [
          Object {
            "_id": "5ba1dd61809c892f580240ed",
            "user": "5b75f023f1957052e52106f5",
          },
          Object {
            "_id": "5ba80533058247301e362bf9",
            "user": "5ba1acfc13b542138b29af68",
          },
          Object {
            "_id": "5ba847153c823e00161763b5",
            "user": "5ba83562ed730446b54def34",
          },
        ]
      }
      text="Tas nisl commlor sit amet, consec!"
      user={
        Object {
          "_id": "5b9eedd621194c102c5280e8",
          "avatar": "//www.gravatar.com/avatar/e13743a7f1db7f4246badd6fd6ff54ff?s=200&r=pg&d=mm",
          "name": "John Doe",
        }
      }
    />
  </div>
</Fragment>
`;
